connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java

```diff
@@ -309,12 +337,17 @@ public boolean startTask(
         final WorkerTask workerTask;
         try {
             final ConnectorConfig connConfig = new ConnectorConfig(connProps);
+            String connType = connConfig.getString(ConnectorConfig.CONNECTOR_CLASS_CONFIG);
+            ClassLoader connectorLoader = plugins.delegatingLoader().connectorLoader(connType);
+            ClassLoader save = Plugins.compareAndSwapLoaders(connectorLoader);
             final TaskConfig taskConfig = new TaskConfig(taskProps);
-
             final Class<? extends Task> taskClass = taskConfig.getClass(TaskConfig.TASK_CLASS_CONFIG).asSubclass(Task.class);
-            final Task task = connectorFactory.newTask(taskClass);
+            final Task task = plugins.newTask(taskClass);
             log.info("Instantiated task {} with version {} of type {}", id, task.version(), taskClass.getName());
 
+            // By maintaining connector's specific class loader for this thread here, we first
+            // search for converters within the connector dependencies, and if not found the
+            // plugin class loader delegates loading to the delegating classloader.
```

Comment:
Every call to `getClass` in an `AbstractConfig` ends up as `return Class.forName(trimmed, true, Utils.getContextOrKafkaClassLoader());`, which is good for us. 

Now, what I'm assuming here is that at this point the thread's classloader is not the delegating classloader, but the connector's `PluginClassLoader`. This child first classloader, will search for the converter within its URLs first (meaning the connector's classes and deps). By having the `DelegatingClassLoader` as parent, if it doesn't find it "locally" it will call the delegating classloader to select the converter among all the options (first in the `plugin.path` and lastly in the classpath). 

In order to apply the search with the delegating classloader initially, I would have to do a compareAndSwap first from the connector's loader to the delegating one, load the converter, and then switch back. I'm not convinced this is needed however. The connector overrides the converter and if it include a dependency probably we should use it. Otherwise the delegating classloader will be called soon anyway. 

I feel the above is a reasonable assumption. 
Btw converter isolation is not missing. But both for the converters and the transformations, the delegation to the thread's classloader is used, instead of an explicit call to `newConverter` or `newTransformation` (for the latter I haven't found a use case yet).